sync-core/src/main/java/com/cloudant/http/HttpConnection.java

```diff
@@ -0,0 +1,234 @@
+package com.cloudant.http;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.io.IOUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+
+/**
+ * Created by tomblench on 23/03/15.
+ */
+
+/**
+ * <p>
+ * A wrapper for <code>HttpURLConnection</code>s.
+ * </p>
+ *
+ * <p>
+ * Provides some convenience methods for making requests and sending/receiving data as streams,
+ * strings, or byte arrays.
+ * </p>
+ *
+ * <p>
+ * Typical usage:
+ * </p>
+ *
+ * <pre>
+ * HttpConnection hc = new HttpConnection("POST", "application/json", new URL("http://somewhere"));
+ * hc.setUsername("user");
+ * hc.setPassword("pass");
+ * hc.requestProperties.put("x-some-header", "some-value");
+ * hc.setInput("{\"hello\": \"world\"});
+ * String result = hc.executeToString();
+ * // get the underlying HttpURLConnection if you need to do something a bit more advanced:
+ * int response = hc.getConnection().getResponseCode();
+ * </pre>
+ *
+ * @see java.net.HttpURLConnection
+ */
+public class HttpConnection  {
+
+    private final String requestMethod;
+    private final String contentType;
+    public final URL url;
+
+    // created in executeInternal
+    private HttpURLConnection connection;
+
+    // optional & settable
+    private String username;
+    private String password;
+
+    // set by the various setInput() methods
+    private InputStream input;
+    private long inputLength;
+
+    public final HashMap<String, String> requestProperties;
+
+    public HttpConnection(String requestMethod,
+                                   String contentType,
+                                   URL url) {
+        this.requestMethod = requestMethod;
+        this.contentType = contentType;
+        this.url = url;
+        this.requestProperties = new HashMap<String, String>();
+    }
+
+    /**
+     * Set the String of data to be sent to the server
+     * @param input String of data to be sent to the server
+     */
+    public void setInput(final String input) {
+        this.input = new ByteArrayInputStream(input.getBytes());
+        // input is in bytes, not characters
+        this.inputLength = input.getBytes().length;
+    }
+
+    /**
+     * Set the byte array of data to be sent to the server
+     * @param input byte array of data to be sent to the server
+     */
+    public void setInput(final byte[] input) {
+        this.input = new ByteArrayInputStream(input);
+        this.inputLength = input.length;
+    }
+
+    /**
+     * Set the InputStream of data to be sent to the server
+     * @param input InputStream of data to be sent to the server
+     */
+    public void setInput(InputStream input) {
+        this.input = input;
+    }
```

Comment:
`setInput` is named as a side effect of `HttpURLConnection` having methods `setDoInput` and `getInputStream` but you could argue that's a leaky abstraction. Request and Response Body are probably easier to think about than input and output.